### ISSUE TITLE
Possible race condition in htex scale down

### ISSUE THREAD

**Issue by benclifford** [Parsl/parsl#1885]
**Describe the bug**
I've seen the below test failure as a one-off in `--config local` tests in the CI on several occasions.

My gut feeling is there's a race condition here - either in the test, or in parsl itself. I'm noting this for later investigation.

The scale down behaviour here is relatively new. Tagging @yadudoc who implemented it.

```
parsl/tests/test_scaling/test_scale_down.py F
=================================== FAILURES ===================================
________________________________ test_scale_out ________________________________
    @pytest.mark.local
    def test_scale_out():
        dfk = parsl.dfk()
    
        # Since we have init_blocks = 0, at this point we should have 0 managers
        print("Executor : ", dfk.executors['htex_local'])
        print("Before")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
        assert len(dfk.executors['htex_local'].connected_managers) == 0, "Expected 0 managers at start"
    
        fus = [sleeper(i) for i in [3, 3, 25, 25, 50]]
    
        for i in range(2):
            fus[i].result()
    
        # At this point, since we have 1 task already processed we should have atleast 1 manager
        print("Between")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
>       assert len(dfk.executors['htex_local'].connected_managers) == 5, "Expected 5 managers once tasks are running"
E       AssertionError: Expected 5 managers once tasks are running
E       assert 3 == 5
E        +  where 3 = len([{'active': True, 'block_id': '1', 'idle_duration': 0, 'manager': '884d27969893', ...}, {'active': True, 'block_id': '...486f', ...}, {'active': True, 'block_id': '2', 'idle_duration': 0.0022573471069335938, 'manager': 'a19000850efd', ...}])
E        +    where [{'active': True, 'block_id': '1', 'idle_duration': 0, 'manager': '884d27969893', ...}, {'active': True, 'block_id': '...486f', ...}, {'active': True, 'block_id': '2', 'idle_duration': 0.0022573471069335938, 'manager': 'a19000850efd', ...}] = HighThroughputExecutor(\n    address=None, \n    address_probe_timeout=None, \n    cores_per_worker=1.0, \n    cpu_affinit...e, \n    worker_logdir_root=None, \n    worker_port_range=(54000, 55000), \n    worker_ports=None, \n    working_dir=None\n).connected_managers
parsl/tests/test_scaling/test_scale_down.py:66: AssertionError
```

**To Reproduce**
unlikely - this looks like a low probability race condition to me.

**Expected behavior**
test should pass

**Environment**
CI, PR #1884

**Comment by ZhuozhaoLi** [comment]
this looks like the time gap between tasks (those sleeps and task lengths) are not long enough in some extreme cases.

**Comment by benclifford** [comment]
I've just encountered this twice, in the CI for PR #1886.

@yadudoc can you decide if you want to fix or disable this test?

**Comment by benclifford** [comment]
I've encountered this again in https://travis-ci.com/github/Parsl/parsl/jobs/446712963

**Comment by benclifford** [comment]
I'm still encountering this.

**Comment by benclifford** [comment]
Looks like scale down was broken by #2104 (cc @jrueb ) but because the test was disabled due to this issue, this was not detected. I'll investigate further but my hypothesis is that the new keep alives introduced in #2104 stop managers becoming idle.

**Comment by benclifford** [comment]
crossref #2171 for different scaling in trouble

**Comment by benclifford** [comment]
I'm still encountering this fairly often in CI.

**Comment by benclifford** [comment]
actually, a variation, in test_scale_out, not test_scale_down:

```
=================================== FAILURES ===================================
________________________________ test_scale_out ________________________________

    @pytest.mark.local
    def test_scale_out():
        dfk = parsl.dfk()
    
        # Since we have init_blocks = 0, at this point we should have 0 managers
        print("Executor : ", dfk.executors['htex_local'])
        print("Before")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
        assert len(dfk.executors['htex_local'].connected_managers) == 0, "Expected 0 managers at start"
    
        fus = [sleeper(i) for i in [3, 3, 25, 25, 50]]
    
        for i in range(2):
            fus[i].result()
    
        # At this point, since we have 1 task already processed we should have atleast 1 manager
        print("Between")
        print("Managers   : ", dfk.executors['htex_local'].connected_managers)
        print("Outstanding: \n", dfk.executors['htex_local'].outstanding)
>       assert len(dfk.executors['htex_local'].connected_managers) == 5, "Expected 5 managers once tasks are running"
E       AssertionError: Expected 5 managers once tasks are running
E       assert 4 == 5
```

**Comment by benclifford** [comment]
I'm getting this on almost every PR at the moment.

**Comment by benclifford** [comment]
After a rewrite/simplification in PR #2497, I have still seen a failure. See this log extract and assertion failure:

```
1669742961.223570 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.dataflow.strategy:213 _
general_strategy DEBUG: Strategy case 1: Executor has no active tasks
1669742961.224305 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.dataflow.strategy:223 _
general_strategy DEBUG: Strategy case 1b: Executor has no active tasks, and more (5) than minimum blocks (2)
1669742961.225055 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.dataflow.strategy:234 _
general_strategy DEBUG: Idle time has reached 5s for executor htex_local; scaling in
1669742961.225858 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.executors.high_throughp
ut.executor:673 scale_in DEBUG: Scale in called, blocks=3, block_ids=[]
1669742961.234504 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.executors.high_throughp
ut.executor:689 scale_in DEBUG: Scale in selecting from 5 blocks
1669742961.242426 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.executors.high_throughp
ut.executor:561 _hold_block DEBUG: [HOLD_BLOCK]: Sending hold to manager: 8c08ed6d4274
1669742961.243523 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.executors.high_throughp
ut.executor:530 hold_worker DEBUG: Sent hold request to manager: 8c08ed6d4274
1669742961.244965 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.executors.high_throughp
ut.executor:561 _hold_block DEBUG: [HOLD_BLOCK]: Sending hold to manager: 623723393f64
1669742961.246182 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.executors.high_throughp
ut.executor:530 hold_worker DEBUG: Sent hold request to manager: 623723393f64
1669742961.247449 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.executors.high_throughp
ut.executor:561 _hold_block DEBUG: [HOLD_BLOCK]: Sending hold to manager: 9ddeba307299
1669742961.248433 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.executors.high_throughp
ut.executor:530 hold_worker DEBUG: Sent hold request to manager: 9ddeba307299
1669742961.249230 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.providers.local.local:2
66 cancel DEBUG: Terminating job/proc_id: 7940
1669742961.355093 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.providers.local.local:2
66 cancel DEBUG: Terminating job/proc_id: 7924
1669742961.462752 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.providers.local.local:2
66 cancel DEBUG: Terminating job/proc_id: 7949
1669742961.574582 2022-11-29 17:29:21 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.process_loggers:28 wrap
ped DEBUG: Normal ending for _general_strategy on thread FlowControl-Thread
```

...

```
1669742966.286203 2022-11-29 17:29:26 MainProcess-6889 FlowControl-Thread-140166693508864 parsl.dataflow.strategy:198 _general_strategy DEBUG: Executor htex_local has 0 active tasks, 2/0 running/pending blocks, and 5 connected workers
```

...


```
1669742966.446465 2022-11-29 17:29:26 MainProcess-6889 MainThread-140170034964288 parsl.tests.test_scaling.test_scale_down:75 test_scale_out INFO: asserting 2 managers remain
```

and then an assertion failure there, because:
    assert len(dfk.executors['htex_local'].connected_managers) == 2, "Expected 2 managers when no tasks, lower bound by min_blocks"

`connection_managers` is 5, not 2.

This is maybe a bug in scale down where managers are not removed from the list of ready managers.

**Comment by benclifford** [comment]
not resolved by #2808 - bad robophrasing in commit message

**Comment by benclifford** [comment]
I haven't seen this for a long time, and revisiting it looks like #2808 *does* fix this issue: that PR replaces a sleep that hopes to wait long enough for scale out (but apparently wasn't enough) with an event based barrier that waits for _max_blocks simultaneous tasks to be executing, with all of those tasks blocked on this barrier before ending.

What this issue #1885 is reporting is that there's an assumption that if I launch 5 tasks I will "quickly" get 5 blocks. It's a bit surprising that a delay of 15s was not long enough for that to happen. But PR #2808 moves to a model that waits until the right number of tasks are running, rather than hoping that blocks will arrive quickly.
